ssz/utils.py

```diff
@@ -138,13 +135,41 @@ def hash_layer(child_layer: Sequence[bytes]) -> Tuple[Hash32, ...]:
     return parent_layer
 
 
-def merkleize(chunks: Sequence[Hash32]) -> Hash32:
-    padded_chunks = pad_chunks(chunks)
-    number_of_layers = int(math.log2(len(padded_chunks))) + 1
-
-    layers = take(number_of_layers, iterate(hash_layer, padded_chunks))
-    root, = last(layers)
-    return root
+def merkleize(chunks: Sequence[Hash32], pad_for=1) -> Hash32:
+    chunk_len = len(chunks)
+    chunk_depth = max(chunk_len - 1, 0).bit_length()
+    max_depth = max(chunk_depth, (pad_for - 1).bit_length())
+    tmp_list = [None for _ in range(max_depth + 1)]
+
+    def merge(leaf, leaf_index):
+        node = leaf
+        layer = 0
+        while True:
+            if leaf_index & (1 << layer) == 0:
+                if leaf_index == chunk_len and layer < chunk_depth:
+                    # Keep going if we are complementing the void to the next power of 2
+                    node = hash_eth2(node + zerohashes[layer])
+                else:
+                    break
+            else:
+                node = hash_eth2(tmp_list[layer] + node)
+            layer += 1
+        tmp_list[layer] = node
+
+    # Merge in leaf by leaf.
+    for leaf_index in range(chunk_len):
+        merge(chunks[leaf_index], leaf_index)
+
+    # Complement with 0 if empty, or if not the right power of 2
+    if 1 << chunk_depth != chunk_len:
+        merge(zerohashes[0], chunk_len)
+
+    # The next power of two may be smaller than the ultimate virtual size,
+    # complement with zero-hashes at each depth.
+    for layer in range(chunk_depth, max_depth):
+        tmp_list[layer + 1] = hash_eth2(tmp_list[layer] + zerohashes[layer])
+
+    return tmp_list[max_depth]
```

Comment:
I think `number_of_layers` should not be calculated by the number of `chunks` so it can be used for the sparse tree, like `List`? I still need an extra input parameter (`pad_for`) to indicate the size of the tree.

Reply:
Yes, you're right. In the interest of time I'd be ok with keeping it as it is for now. We might have to revisit the implementation anyway when we're going to implement tree caching.